// lightdata-tools/src/services/companiesService.js
// services/companies_service.js
import { executeQuery } from "../db_functions/execute_querys/execute_query.js";
import { CustomException } from "../classes/custom_exception.js";

export class CompaniesService {
    /**
     * @param {{ redisClient: any, redisKey?: string }} options
     */
    constructor({ redisClient, redisKey }) {
        // EMPRESAS (igual que tu código original)
        this.redisClient = redisClient;
        this.redisKey = redisKey;
        this.cache = {};

        // Caches EN MEMORIA para lo nuevo (sin Redis/TTL):
        this.zonesCache = {};    // { [companyId]: { [did]: zoneObj } }
        this.accountsCache = {}; // { [companyId]: { [senderId]: { didCliente, didCuenta } } }
        this.clientsCache = {};  // { [companyId]: { [did]: { fecha_sincronizacion, did, codigo, nombre } } }
        this.driversCache = {};  // { [companyId]: { [driverDid]: { id, nombre } } }
    }
    async addCompany(company) {
        this.cache[company.did] = company;
        await this.redisClient.set(this.redisKey, JSON.stringify(this.cache));
    }
    // ================= EMPRESAS (sin cambios) =================
    async loadAll() {
        const data = await this.redisClient.get(this.redisKey);
        this.cache = JSON.parse(data) || {};
        return this.cache;
    }

    async getById(companyId) {
        if (!this.cache[companyId] || Object.keys(this.cache).length === 0) {
            await this.loadAll();
        }
        let company = this.cache[companyId];
        if (!company) {
            await this.loadAll();
            company = this.cache[companyId];
        }
        if (!company) {
            throw new CustomException({
                title: "Empresa no encontrada",
                message: `No se encontró la empresa con el ID "${companyId}" después de recargar la cache.`,
                stack: "",
            });
        }
        return company;
    }

    async getByCode(companyCode) {
        if (Object.keys(this.cache).length === 0) {
            await this.loadAll();
        }
        let found = Object.values(this.cache).find(
            (c) => String(c.codigo) === String(companyCode)
        );
        if (!found) {
            await this.loadAll();
            found = Object.values(this.cache).find(
                (c) => String(c.codigo) === String(companyCode)
            );
        }
        if (!found) {
            throw new CustomException({
                title: "Empresa no encontrada",
                message: `No se encontró la empresa con el código "${companyCode}" después de recargar la cache.`,
                stack: "",
            });
        }
        return found;
    }

    clearCache() {
        this.cache = {};
    }

    // ================= ZONAS (usa tu query/shape) =================
    async loadZones({ db, companyId }) {
        if (!this.zonesCache[companyId]) this.zonesCache[companyId] = {};

        const queryZones = "SELECT * FROM envios_zonas";
        const rows = await executeQuery({ db, queryZones });

        rows.forEach((row) => {
            const did = row.did;
            this.zonesCache[companyId][did] = {
                id: row.id,
                id_origen: row.id_origen,
                fecha_sincronizacion: row.fecha_sincronizacion,
                did: row.did,
                codigo: row.codigo,
                nombre: row.nombre,
                codigos: row.codigos,
                dataGeo: row.dataGeo,
            };
        });

        return this.zonesCache[companyId];
    }

    async getZonesByCompany({ db, companyId }) {
        let companyZones = this.zonesCache[companyId];

        if (!companyZones || Object.keys(this.zonesCache).length === 0) {
            await this.loadZones({ db, companyId });
            companyZones = this.zonesCache[companyId];
        }

        return companyZones || {};
    }

    clearZonesCache() {
        this.zonesCache = {};
    }

    // ================= CUENTAS (usa tu query/shape) =================
    async loadAccountsForCompany({ db, companyId }) {
        if (!this.accountsCache[companyId]) this.accountsCache[companyId] = {};

        const query = `
      SELECT did, didCliente, ML_id_vendedor AS senderId
      FROM clientes_cuentas
      WHERE superado = 0
        AND elim = 0
        AND tipoCuenta = 1
        AND ML_id_vendedor != ''
    `;
        const rows = await executeQuery({ db, query });

        rows.forEach((row) => {
            this.accountsCache[companyId][row.senderId] = {
                didCliente: row.didCliente,
                didCuenta: row.did,
            };
        });

        return this.accountsCache[companyId];
    }

    async getAccountBySenderId({ db, companyId, senderId }) {
        if (!this.accountsCache[companyId] || Object.keys(this.accountsCache[companyId]).length === 0) {
            await this.loadAccountsForCompany({ db, companyId });
        }

        let acc = this.accountsCache[companyId][senderId];
        if (!acc) {
            await this.loadAccountsForCompany({ db, companyId });
            acc = this.accountsCache[companyId][senderId];
        }
        // OJO: tu código original no lanzaba excepción acá; respetamos eso.
        return acc; // puede ser undefined si no existe
    }

    clearAccountsCache() {
        this.accountsCache = {};
    }

    // ================= CLIENTES (usa tu query/shape) =================
    async loadClientsForCompany({ db, companyId }) {
        if (!this.clientsCache[companyId]) this.clientsCache[companyId] = {};

        const query = "SELECT * FROM clientes";
        const rows = await executeQuery({ db: db, query });

        rows.forEach((row) => {
            const clientId = row.did;
            this.clientsCache[companyId][clientId] = {
                fecha_sincronizacion: row.fecha_sincronizacion,
                did: row.did,
                codigo: row.codigoVinculacionLogE,
                nombre: row.nombre_fantasia,
            };
        });

        return this.clientsCache[companyId];
    }

    async getClientsByCompany({ db, companyId }) {
        if (!this.clientsCache[companyId] || Object.keys(this.clientsCache[companyId]).length === 0) {
            await this.loadClientsForCompany({ db, companyId });
        }

        const companyClients = this.clientsCache[companyId];

        if (!companyClients || Object.keys(companyClients).length === 0) {
            throw new CustomException({
                title: "Clientes no encontrados",
                message: `No se encontraron clientes para la empresa ${companyId} después de recargar la cache.`,
            });
        }

        return companyClients;
    }

    clearClientsCache() {
        this.clientsCache = {};
    }

    // ================= DRIVERS (usa tu query/shape) =================
    async loadDriversForCompany({ db, companyId }) {
        if (!this.driversCache[companyId]) this.driversCache[companyId] = {};

        const query = `
      SELECT su.did, su.nombre, su.apellido
      FROM sistema_usuarios_accesos AS sua
      INNER JOIN sistema_usuarios AS su ON sua.did = su.did
      WHERE sua.perfil IN (3, 6)
        AND sua.elim = 0
        AND sua.superado = 0
        AND su.elim = 0
        AND su.superado = 0
    `;
        const rows = await executeQuery({ db, query });

        rows.forEach((row) => {
            this.driversCache[companyId][row.did] = {
                id: row.did,
                nombre: row.nombre + " " + row.apellido,
            };
        });

        return this.driversCache[companyId];
    }

    async getDriversByCompany({ db, companyId }) {
        if (!this.driversCache[companyId] || Object.keys(this.driversCache[companyId]).length === 0) {
            await this.loadDriversForCompany({ db, companyId });
        }

        let companyDrivers = this.driversCache[companyId];

        if (!companyDrivers || Object.keys(companyDrivers).length === 0) {
            await this.loadDriversForCompany({ db, companyId });
            companyDrivers = this.driversCache[companyId];
        }

        if (!companyDrivers || Object.keys(companyDrivers).length === 0) {
            throw new CustomException({
                title: "Drivers no encontrados",
                message: `No se encontraron drivers para la empresa ${companyId} después de recargar la cache.`,
                stack: "",
            });
        }

        return companyDrivers;
    }

    clearDriversCache() {
        this.driversCache = {};
    }
}
