// lightdata-tools/src/classes/estados_envio.js
import { CustomException } from "./custom_exception.js";

export class EstadosEnvio {
    static collected = "collected";
    static atProcessingPlant = "atProcessingPlant";
    static atProcessingPlant2 = "atProcessingPlant2";
    static onTheWay = "onTheWay";
    static returningToProcessingPlant = "returningToProcessingPlant";
    static delivered = "delivered";
    static nobody = "nobody";
    static toBeCollected = "toBeCollected";
    static canceled = "canceled";
    static deliveredSecondVisit = "deliveredSecondVisit";
    static nobodySecondVisit = "nobodySecondVisit";
    static onTheWaySecondVisit = "onTheWaySecondVisit";
    static onTheWayRescheduled = "onTheWayRescheduled";
    static notDelivered = "notDelivered";
    static returnedToClient = "returnedToClient";
    static rejectedByClient = "rejectedByClient";
    static onTheWayThirdVisit = "onTheWayThirdVisit";
    static deliveredThirdVisit = "deliveredThirdVisit";
    static nobodyThirdVisit = "nobodyThirdVisit";
    static deliveredPickupPoint = "deliveredPickupPoint";
    static inClientDepot = "inClientDepot";
    static onTheWayToInterior = "onTheWayToInterior";
    static onEnboxDepot = "onEnboxDepot";
    static invalidAddress = "invalidAddress";

    /**
     * Devuelve el valor numérico del estado según la empresa
     */
    static value(state, companyId) {
        const map = {
            default: {
                collected: 0,
                atProcessingPlant: 1,
                onTheWay: 2,
                returningToProcessingPlant: 3,
                delivered: 5,
                nobody: 6,
                toBeCollected: 7,
                canceled: 8,
                deliveredSecondVisit: 9,
                nobodySecondVisit: 10,
                onTheWaySecondVisit: 11,
                onTheWayRescheduled: 12,
                notDelivered: 13,
                returnedToClient: 14,
                rejectedByClient: 15,
            },
            20: {
                collected: 0,
                atProcessingPlant: 1,
                onTheWay: 2,
                returningToProcessingPlant: 3,
                delivered: 5,
                nobody: 6,
                toBeCollected: 7,
                canceled: 8,
                deliveredSecondVisit: 9,
                nobodySecondVisit: 10,
                onTheWaySecondVisit: 11,
                onTheWayRescheduled: 12,
                notDelivered: 13,
                returnedToClient: 14,
                rejectedByClient: 15,
                onTheWayThirdVisit: 16,
                deliveredThirdVisit: 17,
                nobodyThirdVisit: 18,
            },
            55: {
                collected: 0,
                atProcessingPlant: 1,
                onTheWay: 2,
                returningToProcessingPlant: 3,
                delivered: 5,
                nobody: 6,
                toBeCollected: 7,
                canceled: 8,
                deliveredSecondVisit: 9,
                nobodySecondVisit: 10,
                onTheWaySecondVisit: 11,
                onTheWayRescheduled: 12,
                notDelivered: 13,
                returnedToClient: 14,
                rejectedByClient: 15,
                deliveredPickupPoint: 16,
            },
            72: {
                collected: 0,
                atProcessingPlant: 1,
                onTheWay: 2,
                returningToProcessingPlant: 3,
                delivered: 5,
                nobody: 6,
                toBeCollected: 7,
                canceled: 8,
                deliveredSecondVisit: 9,
                nobodySecondVisit: 10,
                onTheWaySecondVisit: 11,
                onTheWayRescheduled: 12,
                notDelivered: 13,
                returnedToClient: 14,
                rejectedByClient: 15,
                atProcessingPlant2: 16,
            },
            97: {
                collected: 0,
                atProcessingPlant: 1,
                onTheWay: 2,
                returningToProcessingPlant: 3,
                delivered: 5,
                nobody: 6,
                toBeCollected: 7,
                canceled: 8,
                deliveredSecondVisit: 9,
                nobodySecondVisit: 10,
                onTheWaySecondVisit: 11,
                onTheWayRescheduled: 12,
                notDelivered: 13,
                returnedToClient: 14,
                rejectedByClient: 15,
                inClientDepot: 16,
                onTheWayToInterior: 17,
                onEnboxDepot: 18,
                invalidAddress: 22,
            },
        };

        const table = map[companyId] || map.default;
        if (!(state in table))
            throw new CustomException({ title: "Estado inválido", message: `Invalid shipment state '${state}' for companyId ${companyId}` });
        return table[state];
    }

    /**
     * Devuelve el estado según el valor numérico y la empresa
     */
    static type(value, companyId) {
        const inverse = {};

        const fillInverse = (map) => {
            Object.entries(map).forEach(([k, v]) => (inverse[v] = k));
        };

        switch (companyId) {
            case 20:
                fillInverse({
                    0: "collected",
                    1: "atProcessingPlant",
                    2: "onTheWay",
                    3: "returningToProcessingPlant",
                    5: "delivered",
                    6: "nobody",
                    7: "toBeCollected",
                    8: "canceled",
                    9: "deliveredSecondVisit",
                    10: "nobodySecondVisit",
                    11: "onTheWaySecondVisit",
                    12: "onTheWayRescheduled",
                    13: "notDelivered",
                    14: "returnedToClient",
                    15: "rejectedByClient",
                    16: "onTheWayThirdVisit",
                    17: "deliveredThirdVisit",
                    18: "nobodyThirdVisit",
                });
                break;
            case 55:
                fillInverse({
                    0: "collected",
                    1: "atProcessingPlant",
                    2: "onTheWay",
                    3: "returningToProcessingPlant",
                    5: "delivered",
                    6: "nobody",
                    7: "toBeCollected",
                    8: "canceled",
                    9: "deliveredSecondVisit",
                    10: "nobodySecondVisit",
                    11: "onTheWaySecondVisit",
                    12: "onTheWayRescheduled",
                    13: "notDelivered",
                    14: "returnedToClient",
                    15: "rejectedByClient",
                    16: "deliveredPickupPoint",
                });
                break;
            case 72:
                fillInverse({
                    0: "collected",
                    1: "atProcessingPlant",
                    2: "onTheWay",
                    3: "returningToProcessingPlant",
                    5: "delivered",
                    6: "nobody",
                    7: "toBeCollected",
                    8: "canceled",
                    9: "deliveredSecondVisit",
                    10: "nobodySecondVisit",
                    11: "onTheWaySecondVisit",
                    12: "onTheWayRescheduled",
                    13: "notDelivered",
                    14: "returnedToClient",
                    15: "rejectedByClient",
                    16: "atProcessingPlant2",
                });
                break;
            case 97:
                fillInverse({
                    0: "collected",
                    1: "atProcessingPlant",
                    2: "onTheWay",
                    3: "returningToProcessingPlant",
                    5: "delivered",
                    6: "nobody",
                    7: "toBeCollected",
                    8: "canceled",
                    9: "deliveredSecondVisit",
                    10: "nobodySecondVisit",
                    11: "onTheWaySecondVisit",
                    12: "onTheWayRescheduled",
                    13: "notDelivered",
                    14: "returnedToClient",
                    15: "rejectedByClient",
                    16: "inClientDepot",
                    17: "onTheWayToInterior",
                    18: "onEnboxDepot",
                    22: "invalidAddress",
                });
                break;
            default:
                fillInverse({
                    0: "collected",
                    1: "atProcessingPlant",
                    2: "onTheWay",
                    3: "returningToProcessingPlant",
                    5: "delivered",
                    6: "nobody",
                    7: "toBeCollected",
                    8: "canceled",
                    9: "deliveredSecondVisit",
                    10: "nobodySecondVisit",
                    11: "onTheWaySecondVisit",
                    12: "onTheWayRescheduled",
                    13: "notDelivered",
                    14: "returnedToClient",
                    15: "rejectedByClient",
                });
        }

        return inverse[value];
    }
}
