// lightdata-tools/src/endpoints/alta_envio_basica.js
import { EstadosEnvio } from "../classes/estados_envio.js";
import { LogisticaConfig } from "../constants/logisticas_config.js";
import { CustomException } from "../classes/custom_exception.js";

export async function altaEnvioBasica({
  urlAltaEnvioMicroservice,
  urlAltaEnvioRedisMicroservice,
  axiosInstance,
  rabbitServiceInstance,
  queueEstadosML,
  company,
  clientId,
  accountId,
  dataQr,
  flex = 0,
  externo = 0,
  userId,
  driverId,
  lote,
  didExterno,
  nombreClienteEnEmpresaDueña,
  empresaDueña,
}) {
  const idshipment = dataQr?.id ?? null;
  const senderid = dataQr?.sender_id ?? null;

  const payload = {
    companyId: company.did,
    clientId,
    accountId,
    dataQr,
    flex,
    externo,
    userId,
    driverId,
    didExterno,
    nombreClienteEnEmpresaDueña,
    empresaDueña,
    lote,
  };
  let did;

  let data;

  try {
    const response = await axiosInstance.post(urlAltaEnvioMicroservice, payload, {
      headers: { "Content-Type": "application/json" },
    });
    data = response.data;
  } catch (err) {
    throw new CustomException({
      title: "Error en alta de envío",
      message: `Fallo en microservicio: ${err.message}`,
    });
  }

  did = data?.data;

  if (!did)
    throw new CustomException({
      title: "Error al crear envío",
      message: "Microservicio no devolvió un ID válido",
    });

  const syncTasks = [];

  const redisSync = axiosInstance
    .post(
      urlAltaEnvioRedisMicroservice,
      {
        idEmpresa: company.did,
        estado: EstadosEnvio.value(EstadosEnvio.collected, company.did),
        did,
        ml_shipment_id: idshipment,
        ml_vendedor_id: senderid,
      },
      { headers: { "Content-Type": "application/json" } }
    );

  syncTasks.push(redisSync);

  if (LogisticaConfig.sendDataToMLOnCollectEnabled(company.did)) {
    const mlSync = rabbitServiceInstance
      .send(queueEstadosML, {
        idEmpresa: company.did,
        did,
        sellerId: senderid,
        shipmentId: idshipment,
      });

    syncTasks.push(mlSync);
  }

  await Promise.allSettled(syncTasks);

  return did;
}
