// lightdata-tools/src/endpoints/send_shipment_state_to_state_microservice.js
import { connect } from 'amqplib';
import crypto from 'crypto';
import { getHoraLocalDePais } from '../functions/get_hora_local_by_pais.js';
import { getFechaLocalDePais } from '../functions/get_fecha_local_by_pais.js';
import { CustomException } from '../classes/custom_exception.js';

let connection = null;
let channel = null;

async function getChannel(rabbitMQURL, queueEstados) {
    if (channel) return channel;

    connection = await connect(rabbitMQURL);
    channel = await connection.createChannel();
    await channel.assertQueue(queueEstados, { durable: true });

    process.on('exit', () => {
        try {
            if (channel) channel.close();
            if (connection) connection.close();
        } catch (e) {
            console.error('Error closing RabbitMQ connection/channel on exit:', e);
        }
    });

    return channel;
}

export async function sendShipmentStateToStateMicroservice({
    axiosInstance,
    serverEstadosUrl,
    queueEstados,
    rabbitMQURL,
    operacion,
    company,
    userId,
    estado,
    shipmentId,
    latitude,
    longitude,
}) {
    const message = {
        didempresa: company.did,
        didenvio: shipmentId,
        estado: estado,
        subestado: null,
        estadoML: null,
        fecha: getHoraLocalDePais(company.pais),
        quien: userId,
        operacion: operacion,
        latitude,
        longitude,
        tkn: generarTokenFechaHoy(company.pais),
    };

    try {
        const ch = await getChannel(rabbitMQURL, queueEstados);
        const sent = ch.sendToQueue(queueEstados, Buffer.from(JSON.stringify(message)), { persistent: true });

        if (!sent) {
            throw new CustomException({
                title: 'Buffer lleno en RabbitMQ',
                message: 'Buffer lleno en RabbitMQ'
            });
        }
    } catch (error) {
        console.error('Error en RabbitMQ:', error);
        try {
            await axiosInstance.post(serverEstadosUrl, message);
        } catch (httpError) {
            throw new CustomException({
                title: 'Error enviando estado de envío',
                message: `Error enviando estado de envío: ${httpError.message}`,
                originalError: httpError
            });
        }
    }
}
export function generarTokenFechaHoy(country) {
    const fechaLocal = getFechaLocalDePais(country);
    const [anio, mes, dia] = fechaLocal.split('-');

    const fechaString = `${dia}${mes}${anio}`;
    const hash = crypto.createHash('sha256').update(fechaString).digest('hex');

    return hash;
}
