// lightdata-tools/src/functions/verify_token.js
// middlewares/verify_token.js
import jwt from "jsonwebtoken";
import { Status } from "../classes/status.js";
import { CustomException } from "../classes/custom_exception.js";

export function verifyToken({ jwtSecret, issuer, audience }) {
    if (!jwtSecret) throw new CustomException({ title: "verifyToken: falta jwtSecret", message: "El secreto JWT es necesario para la verificación del token" });

    return (req, res, next) => {
        const authHeader = req.get("Authorization") || "";
        const token = authHeader.startsWith("Bearer ")
            ? authHeader.slice(7).trim()
            : null;

        if (!token) {
            return res
                .status(Status.unauthorized)
                .json({ title: "Sin token", message: "Token de autorización faltante" });
        }

        try {
            const payload = jwt.verify(token, jwtSecret, {
                issuer: issuer,
                audience: audience,
            });

            req.user = {
                userId: payload.userId || payload.sub,
                companyId: payload.companyId,
                profile: payload.profile,
            };

            return next();
        } catch (err) {
            return res
                .status(Status.unauthorized)
                .json({ title: "Token inválido", message: err.message });
        }
    };
}
